src/aspire/source/xform.py

```diff
@@ -0,0 +1,388 @@
+        :param which: Either 'forward' or 'adjoint'.
+        :return: An Image object as a result of applying forward or adjoint transformation to `im`.
+        """
+        assert which in ('forward', 'adjoint')
+        # Ensure that we will be able to apply all transformers to the image
+        assert self.n_indices >= im.n_images, f'Can process Image object of max depth {self.n_indices}. Got {im.n_images}.'
+
+        im_data = im.asnumpy().copy()
+
+        # For each individual transformation
+        for i in range(self.n_xforms):
+            # Get the indices corresponding to that transformation
+            idx = np.where(self.indices == i)[0]
+            # For the incoming Image object, find out which transformation indices are applicable
+            idx = np.intersect1d(idx, indices)
+            # For the transformation indices we found, find the indices in the Image object that we'll use
+            im_data_indices = np.where(np.isin(indices, idx))[0]
+            # Apply the transformation to the selected indices in the Image object
+            if len(im_data_indices) > 0:
+                fn_handle = getattr(self.unique_xforms[i], which)
+                im_data[:, :, im_data_indices] = fn_handle(Image(im_data[:, :, im_data_indices])).asnumpy()
+        return Image(im_data)
+
+    def _forward(self, im, indices):
+        return self._forward_or_adjoint(im, indices, 'forward')
+
+    def _adjoint(self, im, indices):
+        return self._forward_or_adjoint(im, indices, 'adjoint')
+
+    def downsample(self, resolution):
+        for xform in self.unique_xforms:
+            xform.downsample(resolution)
+        super().downsample(resolution)
+
+
+def _apply_transform(xform, im, indices, adjoint=False):
+    """
+    A simple global function (i.e. not a method) that is capable of being cached by joblib.Memory object's `cache`
+    method.
+    """
+    if not adjoint:
+        logger.info('  Applying ' + str(xform))
+        return xform.forward(im, indices=indices)
+    else:
+        logger.info('  Applying Adjoint ' + str(xform))
+        return xform.adjoint(im, indices=indices)
+
+
+class Pipeline(Xform):
+    """
+    A Pipeline is a Xform made up of individual transformation steps (i.e. multiple Xform objects).
+    The Pipeline, just like any other Xform, can be run in the `forward` or `adjoint` mode.
+
+    In addition to keeping client-side code clean, a major advantage of Pipeline is that individual steps of the pipeline
+    can be cached transparently by the Pipeline, providing significant performance advantages for steps that are performed
+    repeatedly (especially during development while setting up these pipelines) on any Image/Xform pair. This caching
+    uses joblib.Memory object behind the scenes, but is disabled by default.
+    """
+
+    def __init__(self, xforms=None, memory=None):
+        """
+        Initialize a Pipeline with Xform objects.
```

Comment:
Backticks around `Pipeline` and `Xform`.

Reply:
added